PRD-2023: Fix default route for Ubuntu

bogdando opened the pull request:
Note: depends on https://github.com/Mirantis/fuel/pull/641

Assumes never put several interfaces to single L2 broadcast domain, if one of them is dhcp enabled, otherwise:
ISSUE:
For vlan/bond enabled interfaces, multiple or wrong default routes might have been created
on start-up time, thus, preventing desired gateway x.y.z.t would have been
added for default route on its static subinterface(vlan/bond slave) start-up.

F.e. for eth0 auto with dhcp and its eth0.100 subint with gateway x.y.z.t defined:
1) eth0 starts up
2) default route have created from dhcp-defined gateway f.g.h.i via eth0 (or ethZ, if it shares the 
same L2 broadcast domain with eth0)
3) eth0.100 starts up
4) gateway x.y.z.t ignored as well ...

To avoid this, you can add pre-up hooks for every main/slave interface to erase existing
default route (if any) configured via this or any other main interface, but that would be an ugly solution.

---

ISSUE: Ubuntu affected only.
For interface with gateway defined (/etc/network/interfaces.d/ifcfg-fooX.YYY), there is no ensure
it would has been applied as a default route. It will be applied only after node's
reboot or network restart.

SOLUTION:
We should avoid using "heavy" operations like 'service network restart', thus,
we should as well create default route by manually adding it into routing table, after interface'
configuration and down&up operation has been completed.

We should also ensure right apply order as well (L3_if_downup <||> -> Defaultroute[$def_gateway])

bogdando committed:
Ensure default route is purged before gateway-defined subint will be UP

ISSUE: Ubuntu affected only.
Assumes never put several interfaces to single L2 broadcast domain, if one of them is dhcp enabled, otherwise, for vlan/bond enabled interfaces, multiple or wrong default routes might have been created
on start-up time, thus, preventing desired gateway x.y.z.t would have been
added for default route on its static subinterface(vlan/bond slave) start-up.

F.e. for eth0 auto with dhcp and its eth0.100 subint with gateway x.y.z.t defined:
1) eth0 starts up
2) default route have created from dhcp-defined gateway f.g.h.i via eth0 (or ethZ, if it shares the
same L2 broadcast domain with eth0)
3) eth0.100 starts up
4) gateway x.y.z.t ignored as well ...

To avoid this, you can add pre-up hooks for every main/slave interface to erase existing
default route (if any) configured via this or any other main interface, but that would be an ugly solution.

Fixes: #PRD-2023

Signed-off-by: Bogdan Dobrelya <bogdando@mail.ru>

mirantis-jenkins commented:
:+1: Test PASSed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel_syntax_check/339/

bogdando commented:
Don't merge! Its under test

mirantis-jenkins commented:
:+1: Test PASSed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel_syntax_check/352/

mirantis-jenkins commented:
Merged build triggered.

mirantis-jenkins commented:
Merged build started.

mirantis-jenkins commented:
Merged build finished.

mirantis-jenkins commented:
:+1: Test PASSed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel_syntax_check/355/

bogdando committed:
Add manual default route creation for Ubuntu case as well

ISSUE: Ubuntu affected only.
For interface with gateway defined (/etc/network/interfaces.d/ifcfg-fooX.YYY), there is no ensure
it would has been applied as a default route. It will be applied only after node's
reboot or network restart.

SOLUTION:
We should avoid using "heavy" operations like 'service network restart', thus,
we should as well create default route by manually adding it into routing table, after interface'
configuration and down&up operation has been completed.

We should also ensure right apply order as well (L3_if_downup <||> -> Defaultroute[$def_gateway])

Fixes: #PRD-2023
Depends on: #PRD-1716

Signed-off-by: Bogdan Dobrelya <bogdando@mail.ru>

mirantis-jenkins commented:
Merged build triggered.

mirantis-jenkins commented:
Merged build started.

bogdando commented:
tested ubuntu simple nova-network FLATdhcp all defaults
{"release": "3.2", "nailgun_sha": "b83cc7952d96c268e04359f530d91b92162d149c", "ostf_sha": "f3957b3759263edad5f48da2fb367bd20cbb78cc", "astute_sha": "c688fa1d5e513a2d65d3fb43b25a1b1e45337cff", "fuellib_sha": "826bf712b70757d107aebeca184b34d7f2428cb9"}

mirantis-jenkins commented:
Merged build finished.

mirantis-jenkins commented:
:+1: Test PASSed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel_syntax_check/358/

xenolog commented on deployment/puppet/l23network/templates/ipconfig_Debian_dhcp.erb:
```diff
@@ -7,4 +7,5 @@ iface <%= interface %> inet dhcp
-<% end %>
\ No newline at end of file
+<% end %>
+pre-up route del default <%= /^(\w+\d+)\.(\d+)/.match(interface)[1] rescue /^(\w+\d+)/.match(interface)[1] rescue "nullface".to_s %> || true
```
line11 may break network configuration.

-1

bogdando replied:
OK, will remove it, suggesting the fact "never put several interfaces to single L2 broadcast domain, if one of them is dhcp enabled" would be always TRUE

xenolog commented on deployment/puppet/l23network/templates/ipconfig_Debian_manual.erb:
```diff
@@ -8,4 +8,5 @@ down ip l set <%= interface %> down
-<% end %>
\ No newline at end of file
+<% end %>
+pre-up route del default <%= /^(\w+\d+)\.(\d+)/.match(interface)[1] rescue /^(\w+\d+)/.match(interface)[1] rescue "nullface".to_s %> || true
```
line12 may break network configuration.

-1

bogdando replied:
OK, will remove it, suggesting the fact "never put several interfaces to single L2 broadcast domain, if one of them is dhcp enabled" would be always TRUE

xenolog commented on deployment/puppet/l23network/templates/ipconfig_Debian_static.erb:
```diff
@@ -13,6 +13,7 @@ bond-mode <%= @bond_mode %><% if @bond_miimon %>
 bond-miimon <%= @bond_miimon %><% end %><% if @bond_lacp_rate %>
 bond-lacp-rate <%= @bond_lacp_rate %><% end %>
 <% end %>
+pre-up route del default <%= /^(\w+\d+)\.(\d+)/.match(interface)[1] rescue /^(\w+\d+)/.match(interface)[1] rescue "nullface".to_s %> || true
```
line16 may break network configuration.

bogdando replied:
OK, will remove it, suggesting the fact "never put several interfaces to single L2 broadcast domain, if one of them is dhcp enabled" would be always TRUE

bogdando committed:
Fix network_setup checkbase array indexing

Missed indexing causes wrong startup order for interfaces.

Signed-off-by: Bogdan Dobrelya <bogdando@mail.ru>

bogdando commented on deployment/puppet/l23network/manifests/l3/defaultroute.pp:
```diff
@@ -9,7 +9,11 @@
-        fail("Unsupported for ${::osfamily}/${::operatingsystem}!!! Specify gateway directly for network interface.")
+        exec {'Add default route':
+            path    => '/bin:/usr/bin:/sbin:/usr/sbin',
+            command => "route add default gw ${gateway} || true",
```
TODO ip route replace

mirantis-jenkins commented:
Merged build triggered.

mirantis-jenkins commented:
Merged build started.

mirantis-jenkins commented:
Merged build triggered.

mirantis-jenkins commented:
Merged build finished.

mirantis-jenkins commented:
:+1: Test PASSed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel_syntax_check/374/

mirantis-jenkins commented:
Merged build started.

mirantis-jenkins commented:
Merged build finished.

mirantis-jenkins commented:
:+1: Test PASSed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel_syntax_check/375/

bogdando commented on deployment/puppet/l23network/templates/ipconfig_Debian_bondslave.erb:
```diff
@@ -1,3 +1,4 @@
-bond-master <%= @bond_master %>
\ No newline at end of file
+bond-master <%= @bond_master %>
+pre-up route del default <%= /^(\w+\d+)\.(\d+)/.match(interface)[1] rescue /^(\w+\d+)/.match(interface)[1] rescue "nullface".to_s %> || true
```
Will remove it, suggesting the fact "never put several interfaces to single L2 broadcast domain, if one of them is dhcp enabled" would be always TRUE

bogdando committed:
Use ip route replace instead of route add

Signed-off-by: Bogdan Dobrelya <bogdando@mail.ru>

mirantis-jenkins commented:
Merged build triggered.

mirantis-jenkins commented:
Merged build started.

mirantis-jenkins commented:
Merged build finished.

mirantis-jenkins commented:
:+1: Test PASSed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel_syntax_check/380/

bogdando committed:
Revert "Ensure default route is purged before gateway-defined subint will be UP"

This reverts commit 74c87c740bb0c8bc32098c0bdffd284522fefbd3.

mirantis-jenkins commented:
Merged build started.

mirantis-jenkins commented:
Merged build triggered.

mirantis-jenkins commented:
Merged build triggered.

mirantis-jenkins commented:
Merged build finished.

mirantis-jenkins commented:
:+1: Test PASSed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel_syntax_check/382/

mirantis-jenkins commented:
Merged build started.

mirantis-jenkins commented:
Merged build finished.

mirantis-jenkins commented:
:+1: Test PASSed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel_syntax_check/383/

xenolog commented:
+1

aglarendil commented:
+1
